Adds shots to experimental device interface and integrate with QNode (#4388)
PennyLaneAI/pennylane: doc/releases/changelog-dev.md

* When given a callable, `qml.ctrl` now does its custom pre-processing on all queued operators from the callable.
  [(#4370)](https://github.com/PennyLaneAI/pennylane/pull/4370)

> * `qml.interfaces.set_shots` now has null behaviour for the experimental Device and accepts

Reviewer: What does it mean by "null behaviour"?

Author: Basically, it will not error out but it also won't actually do anything. I could probably come up with some better phrasing.

* When given a callable, `qml.ctrl` now does its custom pre-processing on all queued operators from the callable.
  [(#4370)](https://github.com/PennyLaneAI/pennylane/pull/4370)

> * `qml.interfaces.set_shots` now has null behaviour for the experimental Device and accepts
>   a `Shots` object as well as `int`'s and tuples of `int`'s.

Author: ```suggestion
* `qml.interfaces.set_shots` can now accepts shots specified as a `qml.measurements.Shots` object.
  If a `pennylane.devices.experimental.Device` is provided, the `set_shots` will no longer raise an error, but will simply return without doing anything.
  [(#4388)](https://github.com/PennyLaneAI/pennylane/pull/4388)
```

* When given a callable, `qml.ctrl` now does its custom pre-processing on all queued operators from the callable.
  [(#4370)](https://github.com/PennyLaneAI/pennylane/pull/4370)

* `qml.interfaces.set_shots` can now accepts shots specified as a `qml.measurements.Shots` object.

>   If a `pennylane.devices.experimental.Device` is provided, the `set_shots` will no longer raise an error, but will simply return without doing anything.

Author: This makes a lot of internal code eaiser, like for the original code:
```
                if isinstance(self._original_device, Device):
                     set_shots(self._original_device, override_shots)(self._update_gradient_fn)()
                 else:
                     self._update_gradient_fn(shots=override_shots)
```
We can reduce it to only:
```
                     set_shots(self._original_device, override_shots)(self._update_gradient_fn)()
```
by moving that `isinstance` check inside `set_shots`.

>   If a `pennylane.devices.experimental.Device` is provided, the `set_shots` will no longer raise an error, but will simply return without doing anything.

Author: I guess it comes down to whether or not `set_shots` is an internal tool or user facing.

>   If a `pennylane.devices.experimental.Device` is provided, the `set_shots` will no longer raise an error, but will simply return without doing anything.

Reviewer 2: Ah... @trbromley this is worth double checking/considering. See if people are using `set_shots`, also see if it's been mentioned on github issues, in the forum, etc.

I have definitely been using it.

>   If a `pennylane.devices.experimental.Device` is provided, the `set_shots` will no longer raise an error, but will simply return without doing anything.

Author: Since both Tom and Matthew had concerned, it now raises an explicit `ValueError` with the new device.  Makes some code a little more branch-y, but I'm ok with that.

>   If a `pennylane.devices.experimental.Device` is provided, the `set_shots` will no longer raise an error, but will simply return without doing anything.

Reviewer 3: @albi3ro you're still doing the `if-else` [on this branch](https://github.com/PennyLaneAI/pennylane/blob/shots-on-new-device/pennylane/qnode.py#L944-L949), is that intentional/necessary?

* When given a callable, `qml.ctrl` now does its custom pre-processing on all queued operators from the callable.
  [(#4370)](https://github.com/PennyLaneAI/pennylane/pull/4370)

> * `qml.interfaces.set_shots` accepts `Shots` object as well as `int`'s and tuples of `int`'s.

Reviewer 3: should we add the valueerror with new devices to the changelog? i guess in the breaking changes section?

Author: Before it was an unhelpful `AttributeError`. So not that much of a change.

* When given a callable, `qml.ctrl` now does its custom pre-processing on all queued operators from the callable.
  [(#4370)](https://github.com/PennyLaneAI/pennylane/pull/4370)

> <<<<<<< HEAD
> * `qml.interfaces.set_shots` accepts `Shots` object as well as `int`'s and tuples of `int`'s.
> =======

Author: ```suggestion

* `qml.interfaces.set_shots` accepts `Shots` object as well as `int`'s and tuples of `int`'s.
  [(#4388)](https://github.com/PennyLaneAI/pennylane/pull/4388)

```

* `qml.interfaces.set_shots` can now accept shots specified as a `qml.measurements.Shots` object.
  If a `pennylane.devices.experimental.Device` is provided, the `set_shots` will no longer raise an error, but will simply return without doing anything.

> >>>>>>> 2e776ac7ba5a71dcbab599828607563c08794333

Author: ```suggestion
```

> * `qml.interfaces.set_shots` can now accept shots specified as a `qml.measurements.Shots` object.
>   If a `pennylane.devices.experimental.Device` is provided, the `set_shots` will no longer raise an error, but will simply return without doing anything.
>   [(#4388)](https://github.com/PennyLaneAI/pennylane/pull/4388)

Author: ```suggestion
```
